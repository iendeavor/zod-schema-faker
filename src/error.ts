export class ZodSchemaFakerError extends Error {}
