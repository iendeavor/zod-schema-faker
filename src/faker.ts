import { Faker, faker as _faker } from '@faker-js/faker'
import RandExp from 'randexp'

let _seedValue: number | number[] | undefined
let _shouldSeed: boolean = false

// https://github.com/faker-js/faker/issues/448
// TODO: create standalone faker instead of use following workaround
export const runFake = <Runner extends (faker: Faker) => any>(runner: Runner): ReturnType<Runner> => {
  let oldSeedValue
  if (_shouldSeed) {
    oldSeedValue = _faker.seedValue
    _faker.seed(_seedValue)
  }

  const result = runner(_faker)
  if (result instanceof Promise) {
    throw new SyntaxError('InternalError: runFake cannot be used with async functions')
  }

  if (_shouldSeed) {
    _faker.seed(oldSeedValue)
  }

  _shouldSeed = false

  return result
}

export const randexp = (pattern: string | RegExp, flags?: string) => {
  const randexp = new RandExp(pattern, flags)
  randexp.randInt = (from, to) => runFake(faker => faker.datatype.number({ min: from, max: to }))
  return randexp.gen()
}

/**
 * sets the seed to use.
 */
export const seed = (value?: number | number[]): void => {
  _seedValue = value
  _shouldSeed = true
}
