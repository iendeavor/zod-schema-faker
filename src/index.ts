export { fake } from './fake'
export { seed } from './faker'
export { install, uninstall } from './zod-type-kind-to-zod-type-faker'
export { zodAnyFaker, ZodAnyFaker } from './zod-any-faker'
export { zodArrayFaker, ZodArrayFaker } from './zod-array-faker'
export { zodBigIntFaker, ZodBigIntFaker } from './zod-bigint-faker'
export { zodBooleanFaker, ZodBooleanFaker } from './zod-boolean-faker'
export { zodDateFaker, ZodDateFaker } from './zod-date-faker'
export { zodDefaultFaker, ZodDefaultFaker } from './zod-default-faker'
export { zodDiscriminatedUnionFaker, ZodDiscriminatedUnionFaker } from './zod-discriminated-union-faker'
export { zodEffectsFaker, ZodEffectsFaker } from './zod-effects-faker'
export { zodEnumFaker, ZodEnumFaker } from './zod-enum-faker'
// export { zodFunctionFaker, ZodFunctionFaker } from './zod-function-faker'
export { zodIntersectionFaker, ZodIntersectionFaker } from './zod-intersection-faker'
export { zodLazyFaker, ZodLazyFaker } from './zod-lazy-faker'
export { zodLiteralFaker, ZodLiteralFaker } from './zod-literal-faker'
export { zodMapFaker, ZodMapFaker } from './zod-map-faker'
export { zodNaNFaker, ZodNaNFaker } from './zod-nan-faker'
export { zodNativeEnumFaker, ZodNativeEnumFaker } from './zod-native-enum-faker'
export { zodNeverFaker, ZodNeverFaker } from './zod-never-faker'
export { zodNullFaker, ZodNullFaker } from './zod-null-faker'
export { zodNullableFaker, ZodNullableFaker } from './zod-nullable-faker'
export { zodNumberFaker, ZodNumberFaker } from './zod-number-faker'
export { zodObjectFaker, ZodObjectFaker } from './zod-object-faker'
export { zodOptionalFaker, ZodOptionalFaker } from './zod-optional-faker'
export { zodPromiseFaker, ZodPromiseFaker } from './zod-promise-faker'
export { zodRecordFaker, ZodRecordFaker } from './zod-record-faker'
export { zodSetFaker, ZodSetFaker } from './zod-set-faker'
export { zodStringFaker, ZodStringFaker } from './zod-string-faker'
export { zodTupleFaker, ZodTupleFaker } from './zod-tuple-faker'
export { ZodTypeFaker } from './zod-type-faker'
export { zodUndefinedFaker, ZodUndefinedFaker } from './zod-undefined-faker'
export { zodUnionFaker, ZodUnionFaker } from './zod-union-faker'
export { zodUnknownFaker, ZodUnknownFaker } from './zod-unknown-faker'
export { zodVoidFaker, ZodVoidFaker } from './zod-void-faker'
