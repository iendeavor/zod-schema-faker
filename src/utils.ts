export const assertZodSchema = (schema: any) => {
  if (typeof schema !== 'object' || schema === null || typeof schema._parse !== 'function') {
    throw new Error(`Expected a zod schema, but got ${schema}`)
  }
}
