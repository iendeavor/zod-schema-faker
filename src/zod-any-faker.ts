import * as z from 'zod'
import { ZodTypeFaker } from './zod-type-faker'

export class ZodAnyFaker extends ZodTypeFaker<z.ZodAny> {
  fake(): z.infer<z.ZodAny> {}

  static create(schema: z.ZodAny): ZodAnyFaker {
    return new ZodAnyFaker(schema)
  }
}

export const zodAnyFaker = ZodAnyFaker.create
