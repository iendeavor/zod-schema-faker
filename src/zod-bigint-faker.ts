import * as z from 'zod'
import { runFake } from './faker'
import { ZodTypeFaker } from './zod-type-faker'

export class ZodBigIntFaker extends ZodTypeFaker<z.ZodBigInt> {
  fake(): z.infer<z.ZodBigInt> {
    return runFake(faker => faker.datatype.bigInt())
  }

  static create(schema: z.ZodBigInt): ZodBigIntFaker {
    return new ZodBigIntFaker(schema)
  }
}

export const zodBigIntFaker = ZodBigIntFaker.create
