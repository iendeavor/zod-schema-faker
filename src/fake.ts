import * as z from 'zod'
import { assertZodSchema } from './utils'

export const zodTypeKindToZodTypeFaker: Map<z.ZodFirstPartyTypeKind, any /* TODO: should not use any */> = new Map()

/**
 * generate fake data based on schema
 */
export const fake = <T extends z.ZodType>(schema: T): z.infer<T> => {
  assertZodSchema(schema)

  const typeName = (schema._def as any).typeName
  const faker = zodTypeKindToZodTypeFaker.get(typeName)
  if (faker === undefined) {
    throw ReferenceError(`Unsupported schema: ${typeName}`)
  }

  return faker.create(schema).fake()
}
