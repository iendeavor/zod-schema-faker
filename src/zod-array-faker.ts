import * as z from 'zod'
import { fake } from './fake'
import { runFake } from './faker'
import { ZodTypeFaker } from './zod-type-faker'

export class ZodArrayFaker<T extends z.ZodArray<any, any>> extends ZodTypeFaker<T> {
  fake(): z.infer<T> {
    let min = this.schema._def.minLength?.value ?? this.schema._def.exactLength?.value ?? 0
    let max =
      this.schema._def.maxLength?.value ??
      this.schema._def.exactLength?.value ??
      runFake(faker => faker.datatype.number({ min, max: min + 2 }))

    return Array(runFake(faker => faker.datatype.number({ min, max })))
      .fill(null)
      .map(() => fake(this.schema._def.type))
  }

  static create<T extends z.ZodArray<any, any>>(schema: T): ZodArrayFaker<T> {
    return new ZodArrayFaker(schema)
  }
}

export const zodArrayFaker = ZodArrayFaker.create
