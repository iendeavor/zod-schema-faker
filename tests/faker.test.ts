import * as z from 'zod'
import { install, fake, seed } from '../src'
import { runFake } from '../src/faker'

test('seed', () => {
  install()
  const schema = z.number()

  seed(3)
  const data1 = fake(schema)
  seed(3)
  const data2 = fake(schema)
  expect(data1).toBe(data2)

  const data3 = fake(schema)
  expect(data1).not.toBe(data3)
})

test('runFake can not be used with async functions', () => {
  expect(() => runFake(async () => {})).toThrow()
})
