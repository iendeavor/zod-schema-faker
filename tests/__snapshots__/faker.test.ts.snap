// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seed 1`] = `54543834.398404226`;
